### PR TITLE
Issue 481 add handling derivatives of derived variables

### LINKED ISSUE #481: Add handling of derivatives of derived variables
In SPMe we have explicit expressions for `phi_s`. We want to take gradients of `phi_s` to obtain the ohmic heating. However `phi_s` is not in the boundary conditions (and shouldn't be) so when we take the gradient the resulting symbol is of the wrong size.

### PR DESCRIPTION
**Pull request by Scottmar93**
# Description
Added derivatives of derived variables. Gradients can now be taken of phi_e_n etc for SPMe. Additionally, taking gradients of subdomain variables has been extended to the 2+1D models. Ohmic heating terms in all the thermal models have also been added. 

Still failing 1 debug mode test for 2+1 models. 

Fixes #481 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# Key checklist:

- [x] No style issues: `$ flake8`
- [x] All tests pass: `$ python run-tests.py --unit`
- [ ] Debug mode tests work. 
- [x] The documentation builds: `$ cd docs` and then `$ make clean; make html`

You can run all three at once, using `$ python run-tests.py --quick`.

## Further checks: 

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: pybamm/models/submodels/electrolyte/stefan_maxwell/conductivity/surface_potential_form/leading_surface_form_stefan_maxwell_conductivity.py [hunk 1]
```diff
@@ -62,6 +63,18 @@ def get_coupled_variables(self, variables):
         else:
             return variables
 
+    # def set_boundary_conditions(self, variables):
+    #     pass
```
Comment (rtimms) [inline pybamm/models/submodels/electrolyte/stefan_maxwell/conductivity/surface_potential_form/leading_surface_form_stefan_maxwell_conductivity.py]: can this be removed now?

File: tests/integration/test_models/test_full_battery_models/test_lithium_ion/test_dfn.py [hunk 2]
```diff
@@ -8,6 +8,9 @@
 import unittest
 
 
+pybamm.settings.debug_mode = True
```
Comment (tinosulzer) [inline tests/integration/test_models/test_full_battery_models/test_lithium_ion/test_dfn.py]: should this be removed?

File: tests/unit/test_models/test_submodels/test_electrolyte/test_stefan_maxwell/test_conductivity/test_base_stefan_maxwell_conductivity.py [hunk 3]
```diff
@@ -7,11 +7,16 @@
 import unittest
 
 
+pybamm.settings.debug_mode = True
```
Comment (tinosulzer) [inline tests/unit/test_models/test_submodels/test_electrolyte/test_stefan_maxwell/test_conductivity/test_base_stefan_maxwell_conductivity.py]: should this be removed?

**Comment by Scottmar93** [comment]
> looks great, thanks @Scottmar93 ! maybe not related to this issue, but can we take gradients etc. on variables after solving, and then process the result? or do you have to put them in as variables pre-solve?

Will need to take gradient before and store them in the variables dict at the moment. To simulate the functionality you describe, we could add an option in the `grad` function where if you pass in a processed variable, so that it looks in the varaibles dictionary for the gradient (or we could change the processed variables class so that the gradient is stored within the processed variable class and we can do therefore just do `processed_variable.grad`.

**Pull request merged**